// react-responsiveness/dist/react-responsiveness.js
import n from "react";
const C = {
  Bootstrap_3: {
    xs: 0,
    sm: 576,
    lg: 992
  },
  Bootstrap_4: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  },
  Bootstrap_5: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200,
    xxl: 1400
  },
  Bulma: {
    mobile: 0,
    tablet: 769,
    desktop: 1024,
    widescreen: 1216,
    fullhd: 1408
  },
  Chakra: {
    xs: 0,
    sm: 480,
    md: 768,
    lg: 992,
    xl: 1280,
    "2xl": 1472
  },
  Foundation: {
    small: 0,
    medium: 640,
    large: 1024,
    xlarge: 1200,
    xxlarge: 1440
  },
  Ionic: {
    xs: 0,
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200
  },
  Material_Design: {
    xs: 0,
    sm_8: 600,
    sm_12: 905,
    md: 1240,
    lg: 1440
  },
  Materialize: {
    s: 0,
    m: 600,
    l: 992,
    xl: 1200
  },
  Material_UI: {
    xs: 0,
    sm: 600,
    md: 960,
    lg: 1260,
    xl: 1920
  },
  Quasar: {
    xs: 0,
    sm: 600,
    md: 1024,
    lg: 1440,
    xl: 1920
  },
  Semantic_UI: {
    mobile: 0,
    tablet: 768,
    small_monitor: 992,
    large_monitor: 1200
  },
  Skeleton: {
    mobile: 0,
    phablet: 400,
    tablet: 550,
    desktop: 750,
    desktop_hd: 1e3,
    desktop_hd_lg: 1200
  },
  Tailwind_CSS: {
    xs: 0,
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
    "2xl": 1536
  },
  Windi_CSS: {
    xs: 0,
    sm: 640,
    md: 768,
    lg: 1024,
    xl: 1280,
    "2xl": 1536
  }
}, _ = n.createContext(
  {}
), j = () => n.useContext(_), y = ({
  children: h,
  breakpoints: x = C.Bootstrap_5
}) => {
  const [d, p] = n.useState(""), u = n.useMemo(
    () => Object.entries(x).sort(([, e], [, s]) => (e || 0) - (s || 0)).reduce((e, [s, t], a, m) => {
      var o;
      return e[s] = {
        min: t ? `(min-width: ${t}px)` : "",
        max: (o = m[a + 1]) != null && o[1] ? `(max-width: ${m[a + 1][1] - 0.1}px)` : ""
      }, e;
    }, {}),
    [x]
  ), [l, b] = n.useState(
    Object.assign(
      {},
      ...Object.keys(u).map((e) => ({
        [e]: { min: !1, max: !1, only: !1 }
      }))
    )
  );
  n.useLayoutEffect(() => {
    typeof window < "u" && Object.entries(u).forEach(([e, s]) => {
      const t = {
        min: window.matchMedia(s.min),
        max: window.matchMedia(s.max)
      };
      Object.entries(t).forEach(([a, m]) => {
        const o = ({ matches: g }) => b((i) => {
          var f;
          if (((f = i[e]) == null ? void 0 : f[a]) === g)
            return i;
          const { min: r, max: c } = {
            ...i[e],
            [a]: g
          };
          return e !== d && r && c && p(e), {
            ...i,
            [e]: { min: r, max: c, only: r && c }
          };
        });
        m.addEventListener("change", o), o(m);
      });
    });
  }, [x]);
  const w = (e) => {
    var s;
    return ((s = l[e]) == null ? void 0 : s.min) || !1;
  }, M = (e) => {
    var s;
    return ((s = l[e]) == null ? void 0 : s.min) || !1;
  }, S = (e) => {
    var s, t;
    return ((s = l[e]) == null ? void 0 : s.min) && ((t = l[e]) == null ? void 0 : t.max) || !1;
  };
  return /* @__PURE__ */ n.createElement(
    _.Provider,
    {
      value: { isMin: w, isMax: M, isOnly: S, matches: l, currentInterval: d }
    },
    h
  );
};
export {
  C as Presets,
  y as ResponsivenessProvider,
  j as useResponsiveness
};
